### OBJECTIVE

Convert the feature-engineered dataset into a format that machine learning algorithms can understand while following industry best practices.

In [4]:
import pandas as pd 
import numpy as np 

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import ( OneHotEncoder, StandardScaler )

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [7]:
df = pd.read_csv("../data/featured/telco_feature_engineered.csv")

In [8]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TenureGroup,TotalServices,HasSecurityBundle,AverageMonthlySpend,IsFamilyCustomer
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,Yes,Electronic check,29.85,29.85,No,New Customers,2,1,29.850000,1
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,No,Mailed check,56.95,1889.50,No,2-4 Years,4,2,55.573529,0
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,Yes,Mailed check,53.85,108.15,Yes,New Customers,4,2,54.075000,0
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,No,Bank transfer (automatic),42.30,1840.75,No,2-4 Years,4,3,40.905556,0
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,Yes,Electronic check,70.70,151.65,Yes,New Customers,2,0,75.825000,0


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   customerID           7043 non-null   object 
 1   gender               7043 non-null   object 
 2   SeniorCitizen        7043 non-null   int64  
 3   Partner              7043 non-null   object 
 4   Dependents           7043 non-null   object 
 5   tenure               7043 non-null   int64  
 6   PhoneService         7043 non-null   object 
 7   MultipleLines        7043 non-null   object 
 8   InternetService      7043 non-null   object 
 9   OnlineSecurity       7043 non-null   object 
 10  OnlineBackup         7043 non-null   object 
 11  DeviceProtection     7043 non-null   object 
 12  TechSupport          7043 non-null   object 
 13  StreamingTV          7043 non-null   object 
 14  StreamingMovies      7043 non-null   object 
 15  Contract             7043 non-null   o

In [11]:
df.shape

(7043, 26)

In [12]:
df.isnull().sum()

customerID             0
gender                 0
SeniorCitizen          0
Partner                0
Dependents             0
tenure                 0
PhoneService           0
MultipleLines          0
InternetService        0
OnlineSecurity         0
OnlineBackup           0
DeviceProtection       0
TechSupport            0
StreamingTV            0
StreamingMovies        0
Contract               0
PaperlessBilling       0
PaymentMethod          0
MonthlyCharges         0
TotalCharges           0
Churn                  0
TenureGroup            0
TotalServices          0
HasSecurityBundle      0
AverageMonthlySpend    0
IsFamilyCustomer       0
dtype: int64

In [14]:
df.duplicated().sum()

0

In [68]:
x = df.drop(columns = 'Churn')
y = df['Churn']

In [69]:
x.shape

(7043, 25)

In [70]:
y.shape

(7043,)

In [71]:
x.columns.tolist()

['customerID',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'MonthlyCharges',
 'TotalCharges',
 'TenureGroup',
 'TotalServices',
 'HasSecurityBundle',
 'AverageMonthlySpend',
 'IsFamilyCustomer']

In [72]:
drop_cols = ['customerID']
x = x.drop(columns= drop_cols)

In [73]:
x.shape

(7043, 24)

In [74]:
X_train, X_test, y_train, y_test  = train_test_split(
    x,
    y,
    test_size=0.2, # 20% training data splitted
    random_state = 42,
    stratify=y
)

In [75]:
print("X_train:", X_train.shape)
print("X_test :", X_test.shape)
print("y_train:", y_train.shape)
print("y_test :", y_test.shape)

X_train: (5634, 24)
X_test : (1409, 24)
y_train: (5634,)
y_test : (1409,)


In [76]:
print("Original:")
print(y.value_counts(normalize=True) * 100)

print("\nTraining:")
print(y_train.value_counts(normalize=True) * 100)

print("\nTesting:")
print(y_test.value_counts(normalize=True) * 100)

Original:
No     73.463013
Yes    26.536987
Name: Churn, dtype: float64

Training:
No     73.464679
Yes    26.535321
Name: Churn, dtype: float64

Testing:
No     73.456352
Yes    26.543648
Name: Churn, dtype: float64


### Identify Feature Types

In [77]:
X_test.dtypes

gender                  object
SeniorCitizen            int64
Partner                 object
Dependents              object
tenure                   int64
PhoneService            object
MultipleLines           object
InternetService         object
OnlineSecurity          object
OnlineBackup            object
DeviceProtection        object
TechSupport             object
StreamingTV             object
StreamingMovies         object
Contract                object
PaperlessBilling        object
PaymentMethod           object
MonthlyCharges         float64
TotalCharges           float64
TenureGroup             object
TotalServices            int64
HasSecurityBundle        int64
AverageMonthlySpend    float64
IsFamilyCustomer         int64
dtype: object

In [78]:
numerical_features = ['SeniorCitizen', 'tenure','MonthlyCharges', 'TotalCharges', 'TotalServices','HasSecurityBundle','AverageMonthlySpend']

In [79]:
categorical_features = ["gender",
    "Partner",
    "Dependents",
    "PhoneService",
    "MultipleLines",
    "InternetService",
    "OnlineSecurity",
    "OnlineBackup",
    "DeviceProtection",
    "TechSupport",
    "StreamingTV",
    "StreamingMovies",
    "Contract",
    "PaperlessBilling",
    "PaymentMethod",
    "TenureGroup",
    "IsFamilyCustomer"]

In [80]:
X_train.columns.tolist()

['gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'MonthlyCharges',
 'TotalCharges',
 'TenureGroup',
 'TotalServices',
 'HasSecurityBundle',
 'AverageMonthlySpend',
 'IsFamilyCustomer']

### Create the Transformer

In [81]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import ( StandardScaler, OneHotEncoder)

preprocessor = ColumnTransformer(
    transformers=[
        (
            "num",
            StandardScaler(),
            numerical_features
        ),
        (
            "cat",
            OneHotEncoder( handle_unknown="ignore"),
            categorical_features
        )
    ]
)

### Fit and Transform

In [82]:
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

In [83]:
print(type(X_train_processed))
print(X_train_processed.shape)
print(X_test_processed.shape)

<class 'numpy.ndarray'>
(5634, 54)
(1409, 54)


In [84]:
feature_names = preprocessor.get_feature_names_out()
print(len(feature_names))

54


In [85]:
print(feature_names[:20])

['num__SeniorCitizen' 'num__tenure' 'num__MonthlyCharges'
 'num__TotalCharges' 'num__TotalServices' 'num__HasSecurityBundle'
 'num__AverageMonthlySpend' 'cat__gender_Female' 'cat__gender_Male'
 'cat__Partner_No' 'cat__Partner_Yes' 'cat__Dependents_No'
 'cat__Dependents_Yes' 'cat__PhoneService_No' 'cat__PhoneService_Yes'
 'cat__MultipleLines_No' 'cat__MultipleLines_No phone service'
 'cat__MultipleLines_Yes' 'cat__InternetService_DSL'
 'cat__InternetService_Fiber optic']


### First Pipeline

In [86]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

In [87]:
pipeline = Pipeline(
    steps=[
        ("preprocessor",  preprocessor),
        ("Classifier", LogisticRegression(random_state=42, max_iter=1000))
    ]
)

In [88]:
pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', StandardScaler(),
                                                  ['SeniorCitizen', 'tenure',
                                                   'MonthlyCharges',
                                                   'TotalCharges',
                                                   'TotalServices',
                                                   'HasSecurityBundle',
                                                   'AverageMonthlySpend']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['gender', 'Partner',
                                                   'Dependents', 'PhoneService',
                                                   'MultipleLines',
                                                   'InternetService',
                                                   'OnlineSecurity',
                                                   'OnlineBackup',
                                                   'DeviceProtection',
                                                   'TechSupport', 'StreamingTV',
                                                   'StreamingMovies',
                                                   'Contract',
                                                   'PaperlessBilling',
                                                   'PaymentMethod',
                                                   'TenureGroup',
                                                   'IsFamilyCustomer'])])),
                ('Classifier',
                 LogisticRegression(max_iter=1000, random_state=42))])

In [89]:
prediction = pipeline.predict(X_test)

In [90]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, prediction)
print("Accuracy", accuracy*100)

Accuracy 79.77288857345636


# Notebook Summary

## Objective
The objective of this notebook was to prepare the feature-engineered dataset for machine learning using industry-standard preprocessing techniques while preventing data leakage.

## Tasks Completed

### 1. Loaded Feature Engineered Dataset
- Imported the dataset generated from Notebook 04.
- Verified data integrity before preprocessing.

### 2. Train-Test Split
- Split the dataset into training and testing sets using an 80/20 ratio.
- Used `stratify=y` to preserve the original churn distribution.

### 3. Feature Categorization
- Identified numerical features.
- Identified categorical features.
- Excluded the target variable from preprocessing.

### 4. Numerical Preprocessing
- Applied `StandardScaler` to normalize numerical variables.

### 5. Categorical Preprocessing
- Applied `OneHotEncoder` to categorical variables.
- Used `handle_unknown="ignore"` to safely process unseen categories during prediction.

### 6. ColumnTransformer
- Combined numerical scaling and categorical encoding into a reusable preprocessing pipeline using `ColumnTransformer`.

### 7. Data Transformation
- Fitted the preprocessing pipeline on the training data only.
- Transformed both training and testing datasets using the learned parameters.
- Prevented data leakage by avoiding any fitting on the testing dataset.

### 8. Scikit-Learn Pipeline
- Integrated preprocessing and Logistic Regression into a single Scikit-Learn Pipeline.
- Simplified training and prediction using an end-to-end workflow.

### 9. Baseline Model
- Trained a Logistic Regression model.
- Achieved a baseline accuracy of **79.77%** on the test dataset.

## Key Learning Outcomes

After completing this notebook, the dataset is fully prepared for machine learning.

The following concepts were practiced:

- Train-Test Split
- Data Leakage Prevention
- Feature Categorization
- Feature Scaling
- One-Hot Encoding
- ColumnTransformer
- Pipeline
- Baseline Model Training

The dataset is now ready for systematic model comparison and evaluation in the next notebook.